app/src/main/java/fr/neamar/kiss/IconsHandler.java

```diff
@@ -181,34 +183,42 @@ private Bitmap loadBitmap(String drawableName) {
     /**
      * Get or generate icon for an app
      */
-    public Drawable getDrawableIconForPackage(ComponentName componentName) {
+    public Drawable getDrawableIconForPackage(final ComponentName componentName) {
         try {
-            // system icons, nothing to do
-            if (iconsPackPackageName.equalsIgnoreCase("default")) {
-                return pm.getActivityIcon(componentName);
-            }
+            Drawable icon = null;
 
-            String drawable = packagesDrawables.get(componentName.toString());
-            if (drawable != null) { //there is a custom icon
-                int id = iconPackres.getIdentifier(drawable, "drawable", iconsPackPackageName);
-                if (id > 0) {
-                    //noinspection deprecation: Resources.getDrawable(int, Theme) requires SDK 21+
-                    return iconPackres.getDrawable(id);
+            // Search first in cache
+            icon = cacheGetDrawable(componentName.toString());
+            if (icon != null) {
+                return icon;
+            }
+            
+            // Do we use a custom theme?
+            if (iconsPackPackageName.equalsIgnoreCase("default")) {
+                icon = pm.getActivityIcon(componentName);
+            } else {
+                String drawableIdentifier = packagesDrawables.get(componentName.toString());
+                if (drawableIdentifier != null) { // There is a custom icon
+                    int id = iconPackres.getIdentifier(drawableIdentifier, "drawable", iconsPackPackageName);
+                    if (id > 0) {
+                        // noinspection deprecation: Resources.getDrawable(int, Theme) requires SDK 21+
+                        icon = iconPackres.getDrawable(id);
```

Comment:
maybe you should add

cacheStoreDrawable(componentName.toString(), icon);
return icon;